lib/wit/dependency.py

```diff
@@ -40,16 +40,21 @@ def resolve_deps(self, wsroot, repo_paths, download, source_map, packages, queue
         subdeps = self.package.get_dependencies()
         log.debug("Dependencies for [{}]: [{}]".format(self.name, subdeps))
         for subdep in subdeps:
-            subdep.load_package(packages, repo_paths)
-            subdep.package.load_repo(wsroot, download, subdep.specified_revision)
+            try:
+                subdep.load(packages, repo_paths, wsroot, download)
+            except BadSource as e:
+                errors.append(e)
+                continue
 
             if subdep.name in source_map:
                 if subdep.package.resolve_source(subdep.source) != source_map[subdep.name]:
-                    log.error("Dependency [{}] has multiple conflicting paths:\n"
-                              "  {}\n"
-                              "  {}\n".format(subdep.name, subdep.source,
-                                              source_map[subdep.name]))
-                    sys.exit(1)
+                    if not packages[subdep.name].repo.has_commit(subdep.specified_revision):
+                        error(("Two dependencies have the same name "
```

Comment:
In general, I'd rather we do all erroring from the same location in `main` where we catch `WitUserError`. Concretely, I think this means turning other `error` calls into throwing `WitUserError`, and just setting the message such that it prints the same message.

Why do I think this is useful? It ensures that we can catch and correct or rethrow errors where possible (rather than `sys.exit` occurring deep in the call stack), it unifies the handling of errors such that we can eventually assign return codes, and it enables unit testing since we can catch the exceptions.

Reply:
I would appreciate opinions on this from @richardxia since he's more experienced in Python than me

Reply:
I agree with not overusing a helper function that calls `sys.exit()`, since that makes it impossible to do any handling of errors.

Reply:
Fixed by https://github.com/sifive/wit/pull/135/commits/be5e8ce9f1eb50e877dd8db153afa738a98a326d